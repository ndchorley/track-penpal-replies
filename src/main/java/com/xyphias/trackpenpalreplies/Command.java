package com.xyphias.trackpenpalreplies;

public sealed class Command permits ListPenpals, Quit {}
