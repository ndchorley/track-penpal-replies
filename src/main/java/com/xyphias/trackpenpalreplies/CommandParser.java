package com.xyphias.trackpenpalreplies;

public class CommandParser {
    public static Command parse(String command) {
        if (command.equals("L")) return new ListPenpals();
        
        return new Quit();
    }
}
