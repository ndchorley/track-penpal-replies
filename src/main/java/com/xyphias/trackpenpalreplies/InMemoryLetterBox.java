package com.xyphias.trackpenpalreplies;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class InMemoryLetterBox implements LetterBox {
    public InMemoryLetterBox() {}

    @Override
    public boolean isEmpty() { return letters.isEmpty(); }

    @Override
    public List<Letter> contents() {
        return Collections.unmodifiableList(letters);
    }

    @Override
    public void add(Letter letter) {
        letters.add(letter);
    }

    private final ArrayList<Letter> letters = new ArrayList<>();
}
