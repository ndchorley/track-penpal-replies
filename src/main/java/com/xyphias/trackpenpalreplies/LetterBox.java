package com.xyphias.trackpenpalreplies;

import java.util.List;

public interface LetterBox {
    boolean isEmpty();

    List<Letter> contents();

    void add(Letter letter);

    void remove(Penpal sender);
}
