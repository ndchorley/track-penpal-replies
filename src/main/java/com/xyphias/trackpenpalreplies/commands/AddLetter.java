package com.xyphias.trackpenpalreplies.commands;

import com.xyphias.trackpenpalreplies.Letter;

public record AddLetter(Letter letter) implements Command {}
