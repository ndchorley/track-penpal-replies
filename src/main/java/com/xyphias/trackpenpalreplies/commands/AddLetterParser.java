package com.xyphias.trackpenpalreplies.commands;

import com.xyphias.trackpenpalreplies.Letter;
import com.xyphias.trackpenpalreplies.LetterBox;
import com.xyphias.trackpenpalreplies.Penpal;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class AddLetterParser {
    private final LetterBox letterBox;

    public AddLetterParser(LetterBox letterBox) {
        this.letterBox = letterBox;
    }

    public Command parse(String input) {
        Matcher matcher = matcherFor(input);

        if (!matcher.matches()) return null;

        return commandFrom(matcher);
    }

    private Command commandFrom(Matcher matcher) {
        String name = matcher.group(1);
        LocalDate receivedOn =
                LocalDate.parse(
                        matcher.group(2),
                        DateTimeFormatter.ofPattern("dd/MM/yyyy")
                );

        return new AddLetter(
                new Letter(new Penpal(name), receivedOn),
                letterBox
        );
    }

    private static Matcher matcherFor(String input) {
        return 
                Pattern
                        .compile("A ([\\w ]+);(\\d{2}/\\d{2}/\\d{4})")
                        .matcher(input);
    }
}
