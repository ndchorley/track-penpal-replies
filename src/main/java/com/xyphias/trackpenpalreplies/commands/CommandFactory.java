package com.xyphias.trackpenpalreplies.commands;

import com.xyphias.trackpenpalreplies.Letter;
import com.xyphias.trackpenpalreplies.LetterBox;
import com.xyphias.trackpenpalreplies.Penpal;
import com.xyphias.trackpenpalreplies.infrastructure.io.OutputWriter;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CommandFactory {
    private LetterBox letterBox;
    private OutputWriter outputWriter;

    public CommandFactory(LetterBox letterBox, OutputWriter outputWriter) {
        this.letterBox = letterBox;
        this.outputWriter = outputWriter;
    }

    public Command createFrom(String input) {
        if (input.equals("L")) return new ListLetters(letterBox, outputWriter);
        
        else if (input.startsWith("A")) {
            Matcher matcher = 
                    Pattern.compile("A ([\\w ]+);(\\d{2}/\\d{2}/\\d{4})")
                            .matcher(input);
            matcher.matches();
            
            String name = matcher.group(1);
            LocalDate receivedOn = 
                    LocalDate.parse(
                            matcher.group(2),
                            DateTimeFormatter.ofPattern("dd/MM/yyyy")
                    );
            
            return new AddLetter(
                    new Letter(new Penpal(name), receivedOn),
                    letterBox
            );
        }

        return new Quit();
    }
}
