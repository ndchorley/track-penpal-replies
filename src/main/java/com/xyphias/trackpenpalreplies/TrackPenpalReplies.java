package com.xyphias.trackpenpalreplies;

import com.xyphias.trackpenpalreplies.commands.Command;
import com.xyphias.trackpenpalreplies.commands.CommandFactory;
import com.xyphias.trackpenpalreplies.commands.Quit;
import com.xyphias.trackpenpalreplies.infrastructure.io.InputReader;
import com.xyphias.trackpenpalreplies.infrastructure.io.OutputWriter;

public class TrackPenpalReplies {
    private final InputReader inputReader;
    private final OutputWriter outputWriter;
    private final LetterBox letterBox;
    private final CommandFactory commandFactory;

    public TrackPenpalReplies(
            LetterBox letterBox, InputReader inputReader, OutputWriter outputWriter, CommandFactory commandFactory
    ) {
        this.inputReader = inputReader;
        this.outputWriter = outputWriter;
        this.letterBox = letterBox;
        this.commandFactory = commandFactory;
    }

    public void run() {
        while (true) {
            Command command = readCommand();

            outputWriter.writeLine("");
            
            if (command instanceof Quit) break;
            
            command.execute();
        }
    }

    private Command readCommand() {
        outputWriter.write("enter command: ");

        String input = inputReader.readLine();

        Command command = commandFactory.createFrom(input);
        
        return command;
    }
}
