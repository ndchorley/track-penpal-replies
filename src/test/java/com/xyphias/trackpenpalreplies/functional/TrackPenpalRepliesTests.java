package com.xyphias.trackpenpalreplies.functional;

import com.xyphias.trackpenpalreplies.LetterBox;
import com.xyphias.trackpenpalreplies.TrackPenpalReplies;
import com.xyphias.trackpenpalreplies.commands.CommandFactory;
import com.xyphias.trackpenpalreplies.fakes.io.CapturingOutputWriter;
import com.xyphias.trackpenpalreplies.fakes.io.InMemoryInputReader;
import com.xyphias.trackpenpalreplies.fakes.db.InMemoryLetterBox;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;

public class TrackPenpalRepliesTests {
    private final InMemoryInputReader inputReader = new InMemoryInputReader();
    private final LetterBox letterBox = new InMemoryLetterBox();
    private final CapturingOutputWriter outputWriter = new CapturingOutputWriter();
    private final CommandFactory commandFactory = new CommandFactory(letterBox, outputWriter);
    private final TrackPenpalReplies app = new TrackPenpalReplies(inputReader, outputWriter, commandFactory);

    @Test
    public void it_displays_a_message_when_no_letters_need_a_reply() {
        List<String> commands = List.of("L", "Q");
        inputReader.withInputs(commands);

        app.run();

        Approvals.verify(outputWriter.written);
    }

    @Test
    public void it_displays_the_details_of_a_letter() {
        List<String> commands = List.of("A Amandine;04/07/2024", "L", "Q");
        inputReader.withInputs(commands);

        app.run();

        Approvals.verify(outputWriter.written);
    }
    
    @Test
    public void it_displays_older_letters_before_newer_ones() {
        List<String> commands = 
                List.of(
                        "A Marta;11/08/2024",
                        "A Amandine;04/07/2024",
                        "A Eric;01/06/2024",
                        "L",
                        "Q"
                );
        inputReader.withInputs(commands);

        app.run();

        Approvals.verify(outputWriter.written);        
    }
    
    @Test
    public void a_letter_can_be_removed() {
        List<String> commands =
                List.of(
                        "A John;07/08/2024",
                        "A Amandine;04/07/2024",
                        "R Amandine",
                        "L",
                        "Q"
                );
        inputReader.withInputs(commands);

        app.run();

        Approvals.verify(outputWriter.written);
    }
}
