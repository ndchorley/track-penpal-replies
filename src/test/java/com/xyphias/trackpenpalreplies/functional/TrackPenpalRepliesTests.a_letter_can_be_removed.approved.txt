enter command: 
enter command: 
enter command: 
enter command: 
John, 07 August 2024
enter command: 
